### PR TITLE
Fixes concurrency issues that happen to tag files when read by multiple threads

### PR DESCRIPTION
**Pull request by samgabriel**
We were experiencing an issue when using parallel compilation where a racing condition was causing tag files to be overwritten and become corrupted thus causing jspc to fail. In order to fix that problem a new class was added that extends the JspC and provides access to the JspCServletContext. Using the JspCServletContext we can reuse the same JspRuntimeContext for each of the Workers. Jasper has a synchronization block on the JspRuntimeContext to ensure tag files are only being processed by one thread at a time. We no longer received this random errors due to tag corruption.

As an added bonus the tld scan also happens once since we are reusing the same scanner.

### CODE REVIEW

**Comment by leonardehrenfried** [comment]
Thanks for the pull request @samgabriel.

This looks good to me and the tests still pass, so I'm happy to merge this in.

I'm going to get @tcollignon to review this as well.

In the meanwhile, if you want to be fantastic you can add a test that compiles a tag file. That is something not currently covered by a test.

File: src/test/java/io/leonard/maven/plugins/jspc/TestJspcMojo.java [hunk 1]
```diff
@@ -34,12 +37,18 @@ public void should_return_webfrag_equal_to_reference_when_executeMojo_on_project
...
     Path webfrag = Paths.get("target/test-classes/unit/project_one_jsp/target/webfrag.xml");
-    assertThat(Files.readAllBytes(webfrag)).isEqualTo(Files.readAllBytes(expectedWebfrag));
+    
+    byte[] actualWebfrag = Files.readAllBytes(webfrag);
+	byte[] expectedWebFrag = Files.readAllBytes(expectedWebfrag);
+//	String actualWebFragStr = new String(actualWebfrag).replaceAll("\n", "\r\n");
+	
+//	String expectedWebFragStr = new String(expectedWebFrag);
+	assertThat(actualWebfrag).isEqualTo(expectedWebFrag);
+//	assertThat(actualWebFragStr).isEqualTo(expectedWebFragStr);
```
Comment (tcollignon) [inline src/test/java/io/leonard/maven/plugins/jspc/TestJspcMojo.java]: why this change? only for avoir one line?

Comment (tcollignon) [inline src/test/java/io/leonard/maven/plugins/jspc/TestJspcMojo.java]: why this? only to avoir one line?

**Comment by tcollignon** [comment]
Hi,

Yes I will take a look at this
And yes I agree to @leonardehrenfried , adding a test with a tag file may be very good

tx @samgabriel !

**Comment by leonardehrenfried** [comment]
@tcollignon What should be do about this PR?

**Comment by tcollignon** [comment]
Yes most important is missing test
I will try to add this myself
I'm interesting in this PR because I have sometimes this problem too

**Comment by leonardehrenfried** [comment]
Ok, I will merge and wait for your test. Thanks!
